app/ui/browser-modern/sagas/page-sagas.js

```diff
@@ -77,6 +81,32 @@ export function* destroyPageSession({ id, withUI }) {
   }
 }
 
+function* forkPage(pageId, historyIndex) {
+  const currentPageIndex = yield select(PageSelectors.getPageIndexById, pageId);
+  const historyURLs = yield select(PageSelectors.getPageHistoryURLs, pageId);
+  const historyList = escape(JSON.stringify(historyURLs));
+  const url = `${SESSION_RESTORE_ADDR}/?history=${historyList}&historyIndex=${historyIndex}`;
+  const action = PageEffects.createPageSession(url, { selected: true });
+  const { id } = action;
+
+  yield call(createPageSession, action);
+  yield put(PageActions.setPageIndex(id, currentPageIndex + 1));
```

Comment:
Reading the code:

```
function setPageIndex(state, pageId, pageIndex) {
  return state.withMutations(mut => {
    const oldIndex = state.displayOrder.findIndex(id => id === pageId);
    mut.update('displayOrder', l => l.delete(oldIndex));
    mut.update('displayOrder', l => l.insert(pageIndex, pageId));
  });
}
```

This looks like it will misbehave if the `pageId` is not found in `displayOrder`: `findIndex` will return `-1`, and the `.delete` will wrap around the negative index and delete the last element.

I think `setPageIndex` should check whether `oldIndex` is `-1`, which avoids doing damage if calling code — like this code — fails to create the new tab.